corehq/form_processor/backends/sql/update_strategy.py

```diff
@@ -287,6 +301,30 @@ def rebuild_from_transactions(self, transactions, rebuild_transaction, unarchive
         if not self.case.modified_on:
             self.case.modified_on = rebuild_transaction.server_date
 
+    def reconcile_transactions_if_necessary(self, xform):
+        if not self.case.check_transaction_order():
+            try:
+                self.reconcile_transactions(rebuild=True)
+            except ReconciliationError:
+                pass
+
+    def reconcile_transactions(self, rebuild=False):
+        transactions = self.case.transactions
+        sorted_transactions = sorted(
+            transactions,
+            key=_transaction_sort_key_function(self.case)
+        )
+        if sorted_transactions:
+            if sorted_transactions[0].type & const.CASE_ACTION_CREATE != sorted_transactions[0].type:
+                error = "Case {0} first transaction not create transaction: {1}"
+                raise ReconciliationError(
+                    error.format(self.case.case_id, sorted_transactions[0])
+                )
+        if rebuild:
+            rebuild_detail = RebuildWithReason(reason="client_date_reconciliation")
+            rebuild_transaction = CommCareCaseSQL.rebuild_transaction(self.case, rebuild_detail)
+            self.rebuild_from_transactions(sorted_transactions, rebuild_transaction)
```

Comment:
Prior to calling this we should bulk fetch the forms and cache them on the transactions rather than fetch the individually.

Currently that's done here: https://github.com/dimagi/commcare-hq/blob/baa7fe2cc8faaabf53b906734ffcbc25a8212c88/corehq/form_processor/backends/sql/processor.py#L351

Could refactor that to make it useful in both instances.

Reply:
70be6e9